File: client-android/serenada-core/src/test/java/app/serenada/core/fakes/FakeMediaEngine.kt
package app.serenada.core.fakes

import android.content.Intent
import app.serenada.core.call.LocalCameraMode
import app.serenada.core.call.PeerConnectionSlotProtocol
import app.serenada.core.call.SessionMediaEngine
import org.webrtc.EglBase
import org.webrtc.IceCandidate
import org.webrtc.PeerConnection
import org.webrtc.RendererCommon
import org.webrtc.SurfaceViewRenderer
import org.webrtc.VideoSink
import org.webrtc.VideoTrack

internal class FakeMediaEngine : SessionMediaEngine {
    var startLocalMediaCalls = 0
        private set
    var releaseCalls = 0
        private set
    val toggleAudioCalls = mutableListOf<Boolean>()
    val toggleVideoCalls = mutableListOf<Boolean>()
    var iceServersSet = false
        private set
    val createdSlotCids = mutableListOf<String>()
    val removedSlots = mutableListOf<PeerConnectionSlotProtocol>()
    val fakeSlots = mutableMapOf<String, FakePeerConnectionSlot>()
    var failNextCreatedSlotRemoteOffer = false
    var deferNextCreatedSlotOfferSdp = false
    var deferNextCreatedSlotAnswerSdp = false

    private var _iceServers: List<PeerConnection.IceServer>? = null

    // Camera capture starts (true = video capture requested). Tracked SEPARATELY
    // from content (screen share) starts so tests can assert camera vs content
    // lifecycle independently in independent mode.
    val startVideoCaptureCalls = mutableListOf<Boolean>()
    // Content (screen share) start/stop counts. In independent mode these are the
    // CONTENT track lifecycle; they never increment camera counts.
    var startScreenShareCalls = 0
        private set
    var stopScreenShareCalls = 0
        private set
    /** Alias for content-start count, to read clearly in independent-mode tests. */
    val contentShareStartCalls: Int get() = startScreenShareCalls
    /** Alias for content-stop count. */
    val contentShareStopCalls: Int get() = stopScreenShareCalls
    // Per-peer capability params seen at slot creation (cid → supported / policy).
    val createdSlotSupportsIndependent = mutableMapOf<String, Boolean>()
    val createdSlotOfferOwner = mutableMapOf<String, Boolean>()
    // Result the fake reports for screen-share start/stop. Defaults to success
    // so session-level content_state/revision flow can be exercised.
    var startScreenShareResult = true
    var stopScreenShareResult = true
    // When true, startScreenShare / stopScreenShare drive the per-peer content
    // attach loop on the modeled fake slots (capable peers only), mirroring the
    // real WebRtcEngine's `peerSlots.forEach { attachLocalTracksToSlot }`. Lets
    // session-level tests observe per-peer attach-failure ISOLATION (a failing
    // peer renegotiates while a healthy peer carries the share) without a real
    // VideoTrack. Off by default so existing tests are unaffected.
    var modelIndependentContentAttach = false

    val attachLocalContentSinkCalls = mutableListOf<VideoSink>()
    val detachLocalContentSinkCalls = mutableListOf<VideoSink>()

    // Hold/resume observation.
    var suspendLocalMediaForHoldCalls = 0
        private set
    /** Recorded (audioEnabled, videoMode) pairs passed to resume. */
    val resumeLocalMediaFromHoldCalls = mutableListOf<Pair<Boolean, LocalCameraMode?>>()
    /** Recorded enabled values passed to setRemotePlaybackEnabled. */
    val setRemotePlaybackEnabledCalls = mutableListOf<Boolean>()
    var detachRenderersForHoldCalls = 0
        private set

    // FIX A1 model: whether the engine currently holds a live mic CAPTURE track.
    // Mirrors the real WebRtcEngine: startLocalMedia creates it, suspend releases
    // it, and resume recreates it ONLY when audio is desired (a muted resume must
    // leave it released so the OS mic indicator stays off). Counts recreations so
    // a test can assert a muted resume does NOT recreate the mic.
    var micCaptureTrackPresent = false
        private set
    var micCaptureRecreateCount = 0
        private set

    // Simulates a failed mic ACQUIRE: when true, an enabling toggleAudio(true)
    // cannot (re)create the capture track and reports the effective state as
    // false (no live track). Mirrors the real engine returning false when the OS
    // denies/loses the mic. Lets a test assert the broadcast/published audio
    // state never reports live audio that isn't actually captured (FIX P5).
    var failMicAcquire = false

    // FIX P5 model: whether the engine currently holds a live camera CAPTURE
    // track. Mirrors the real WebRtcEngine: startLocalMedia/resume(camera) create
    // it, suspend + resume-camera-off release it. A FOREGROUND enable with no
    // track (toggleVideo(true)) must (re)create + attach it before publishing
    // true. Counts recreations so a test can assert the (re)create on enable.
    var cameraCaptureTrackPresent = false
        private set
    var cameraCaptureRecreateCount = 0
        private set

    override fun startLocalMedia(startVideoCapture: Boolean) {
        startLocalMediaCalls++
        startVideoCaptureCalls.add(startVideoCapture)
        micCaptureTrackPresent = true
        cameraCaptureTrackPresent = startVideoCapture
    }

    // Multi-call held join: createSendersForHold creates stable senders WITHOUT
    // capture (contract §5 / Core Invariant 3). Tracked so a held-join test can
    // assert senders were created while NO mic/camera capture track exists.
    var createSendersForHoldCalls = 0
        private set
    override fun createSendersForHold() {
        createSendersForHoldCalls++
        // Senders exist but carry NO capture track (the held invariant): leave
        // micCaptureTrackPresent / cameraCaptureTrackPresent false.
    }

    override fun release() { releaseCalls++ }

    override fun suspendLocalMediaForHold() {
        suspendLocalMediaForHoldCalls++
        micCaptureTrackPresent = false
        cameraCaptureTrackPresent = false
    }
    override fun resumeLocalMediaFromHold(audioEnabled: Boolean, videoMode: LocalCameraMode?) {
        resumeLocalMediaFromHoldCalls.add(audioEnabled to videoMode)
        // Recreate the mic capture ONLY when audio is desired (FIX A1). A muted
        // resume keeps capture released — no recreate, sender track stays null.
        if (audioEnabled && !micCaptureTrackPresent) {
            micCaptureTrackPresent = true
            micCaptureRecreateCount++
        }
        // Recreate the camera capture ONLY when a camera mode is desired. A
        // camera-off resume keeps it released (FIX P5: a later video-on toggle
        // must (re)create it via toggleVideo).
        if (videoMode != null && videoMode != LocalCameraMode.SCREEN_SHARE && !cameraCaptureTrackPresent) {
            cameraCaptureTrackPresent = true
            cameraCaptureRecreateCount++
        }
    }
    override fun setRemotePlaybackEnabled(enabled: Boolean) {
        setRemotePlaybackEnabledCalls.add(enabled)
    }
    override fun detachRenderersForHold() {
        detachRenderersForHoldCalls++
    }
    override fun toggleAudio(enabled: Boolean): Boolean {
        toggleAudioCalls.add(enabled)
        // Simulated acquire failure: an enable cannot obtain a live mic track, so
        // the effective state is false (no recreate, no live track).
        if (enabled && failMicAcquire) {
            micCaptureTrackPresent = false
            return false
        }
        // FIX P5: a FOREGROUND enable with no live mic track (re)creates + attaches
        // it before publishing — mirrors the real engine ensuring the track exists.
        // When the track already exists this is a plain setEnabled (no recreate),
        // preserving single-call mute/unmute behavior. Returns the EFFECTIVE state.
        if (enabled && !micCaptureTrackPresent) {
            micCaptureTrackPresent = true
            micCaptureRecreateCount++
        }
        return enabled && micCaptureTrackPresent
    }
    override fun toggleVideo(enabled: Boolean): Boolean {
        toggleVideoCalls.add(enabled)
        // FIX P5: a FOREGROUND enable with no live camera track (re)creates +
        // attaches it before returning true. When the track already exists this is
        // a plain setEnabled (no recreate). Returns the EFFECTIVE state.
        if (enabled && !cameraCaptureTrackPresent) {
            cameraCaptureTrackPresent = true
            cameraCaptureRecreateCount++
        } else if (!enabled) {
            cameraCaptureTrackPresent = false
        }
        return enabled && cameraCaptureTrackPresent
    }
    var flipCameraCalls = 0
        private set
    override fun flipCamera() { flipCameraCalls++ }

    /**
     * Engine-side active camera mode. Tests set this to simulate the camera being
     * in WORLD/COMPOSITE while an independent screen share runs, so the session's
     * post-stop camera-hint restore can be asserted. Null mirrors a no-camera
     * engine (the session then falls back to its own [LocalCameraMode] copy).
     */
    var activeCameraMode: LocalCameraMode? = null
    override fun activeCameraMode(): LocalCameraMode? = activeCameraMode
    override fun startScreenShare(intent: Intent): Boolean {
        startScreenShareCalls++
        if (startScreenShareResult && modelIndependentContentAttach) {
            // Per-peer attach: only capable slots get the content track on their
            // content sender (mirrors attachLocalTracksToSlot's capable branch).
            // A per-slot failNextContentAttach turns that peer's attach into a
            // reject + renegotiation; other peers are unaffected (isolation).
            fakeSlots.values
                .filter { it.supportsIndependentContentVideo }
                .forEach { it.simulateContentAttach(attach = true) }
        }
        return startScreenShareResult
    }
    override fun stopScreenShare(): Boolean {
        stopScreenShareCalls++
        if (stopScreenShareResult && modelIndependentContentAttach) {
            fakeSlots.values
                .filter { it.supportsIndependentContentVideo }
                .forEach { it.simulateContentAttach(attach = false) }
        }
        return stopScreenShareResult
    }

    override fun setIceServers(servers: List<PeerConnection.IceServer>) {
        _iceServers = servers
        iceServersSet = true
        fakeSlots.values.forEach { it.setIceServers(servers) }
    }

    override fun hasIceServers(): Boolean = _iceServers != null

    override fun createSlot(
        remoteCid: String,
        onLocalIceCandidate: (String, IceCandidate) -> Unit,
        onRemoteVideoTrack: (String, VideoTrack?) -> Unit,
        onConnectionStateChange: (String, PeerConnection.PeerConnectionState) -> Unit,
        onIceConnectionStateChange: (String, PeerConnection.IceConnectionState) -> Unit,
        onSignalingStateChange: (String, PeerConnection.SignalingState) -> Unit,
        onRenegotiationNeeded: (String) -> Unit,
        supportsIndependentContentVideo: Boolean,
        isOfferOwner: () -> Boolean,
    ): PeerConnectionSlotProtocol {
        createdSlotCids.add(remoteCid)
        createdSlotSupportsIndependent[remoteCid] = supportsIndependentContentVideo
        createdSlotOfferOwner[remoteCid] = isOfferOwner()
        val slot = FakePeerConnectionSlot(
            remoteCid = remoteCid,
            onLocalIceCandidate = onLocalIceCandidate,
            onConnectionStateChange = onConnectionStateChange,
            onIceConnectionStateChange = onIceConnectionStateChange,
            onSignalingStateChange = onSignalingStateChange,
            onRenegotiationNeeded = onRenegotiationNeeded,
            supportsIndependentContentVideo = supportsIndependentContentVideo,
        )
        if (failNextCreatedSlotRemoteOffer) {
            slot.failNextRemoteOffer = true
            failNextCreatedSlotRemoteOffer = false
        }
        if (deferNextCreatedSlotOfferSdp) {
            slot.deferNextOfferSdp = true
            deferNextCreatedSlotOfferSdp = false
        }
        if (deferNextCreatedSlotAnswerSdp) {
            slot.deferNextAnswerSdp = true
            deferNextCreatedSlotAnswerSdp = false
        }
        fakeSlots[remoteCid] = slot
        _iceServers?.let(slot::setIceServers)
        return slot
    }

    override fun removeSlot(slot: PeerConnectionSlotProtocol) {
        removedSlots.add(slot)
        fakeSlots.remove(slot.remoteCid)
    }

    val attachLocalSinkCalls = mutableListOf<VideoSink>()
    val detachLocalSinkCalls = mutableListOf<VideoSink>()
    override fun attachLocalRenderer(renderer: SurfaceViewRenderer, rendererEvents: RendererCommon.RendererEvents?) {}
    override fun detachLocalRenderer(renderer: SurfaceViewRenderer) {}
    override fun attachLocalSink(sink: VideoSink) {
        attachLocalSinkCalls += sink
    }
    override fun detachLocalSink(sink: VideoSink) {
        detachLocalSinkCalls += sink
    }
    override fun attachLocalContentSink(sink: VideoSink) {
        attachLocalContentSinkCalls += sink
    }
    override fun detachLocalContentSink(sink: VideoSink) {
        detachLocalContentSinkCalls += sink
    }
    override fun initRenderer(renderer: SurfaceViewRenderer, rendererEvents: RendererCommon.RendererEvents?) {}
    override fun adjustWorldCameraZoom(scaleFactor: Float): Boolean = false
    override fun toggleFlashlight(): Boolean = false
    override fun getEglContext(): EglBase.Context =
        throw UnsupportedOperationException("EGL context not available in tests")

    var nextLocalAudioLevel: Float? = null
    override fun collectLocalAudioLevel(onComplete: (Float?) -> Unit) {
        onComplete(nextLocalAudioLevel)
    }
}

File: client-ios/SerenadaCore/Tests/SerenadaCoreTests/Fakes/FakeMediaEngine.swift
import CoreGraphics
import Foundation
@testable import SerenadaCore
#if canImport(WebRTC)
import WebRTC
#endif

@MainActor
final class FakeMediaEngine: SessionMediaEngine {
    private(set) var startLocalMediaCalls: [Bool] = []
    private(set) var releaseCalls = 0
    private(set) var toggleAudioCalls: [Bool] = []
    private(set) var restartAudioUnitCalls = 0
    private(set) var toggleVideoCalls: [Bool] = []
    private(set) var iceServersSet = false
    private(set) var createdSlotCids: [String] = []
    private(set) var removedSlots: [any PeerConnectionSlotProtocol] = []
    private(set) var fakeSlots: [String: FakePeerConnectionSlot] = [:]
    var failNextCreatedSlotRemoteOffer = false

    private var _iceServers: [IceServerConfig]?
    private var onCameraFacingChanged: ((Bool) -> Void)?
    private var onCameraModeChanged: ((LocalCameraMode) -> Void)?
    private var onFlashlightStateChanged: ((Bool, Bool) -> Void)?
    private var onScreenShareStopped: (() -> Void)?
    private var onZoomFactorChanged: ((Double) -> Void)?
    private var onFeatureDegradation: ((FeatureDegradationState) -> Void)?

    /// Faithful model of local capture-track presence, mirroring WebRtcEngine's
    /// `localAudioTrack`/`localVideoTrack`. Hold releases them; resume reacquires
    /// per desired intent (so a muted/camera-off resume leaves them nil); the
    /// enabling toggles recreate them when missing (the P5 ensure-track fix).
    private(set) var hasLocalAudioTrack = false
    private(set) var hasLocalVideoTrack = false
    /// Count of toggle-driven track RECREATIONS (the resume-then-enable repair),
    /// distinct from the initial acquire / resume acquire. Lets a regression test
    /// assert that a normal unmute with an existing track recreates NOTHING.
    private(set) var toggleAudioTrackRecreations = 0
    private(set) var toggleVideoTrackRecreations = 0

    /// Test seam: simulate a foreground call that has lost its video track (e.g. a
    /// camera-off resume followed by a desired-video-on toggle), so a subsequent
    /// `toggleVideo(true)` must recreate it. Counts as no recreation itself.
    func dropLocalVideoTrackForTesting() { hasLocalVideoTrack = false }

    func startLocalMedia(preferVideo: Bool) {
        startLocalMediaCalls.append(preferVideo)
        // Initial acquire: mic always, camera when preferVideo (mirrors the real
        // engine's startLocalMedia).
        hasLocalAudioTrack = true
        hasLocalVideoTrack = preferVideo
    }

    func release() {
        releaseCalls += 1
        hasLocalAudioTrack = false
        hasLocalVideoTrack = false
    }

    private(set) var suspendLocalMediaForHoldCalls = 0
    private(set) var resumeLocalMediaFromHoldCalls: [(audioEnabled: Bool, videoMode: LocalCameraMode?)] = []
    private(set) var setRemotePlaybackEnabledCalls: [Bool] = []
    private(set) var detachRenderersForHoldCalls = 0

    func suspendLocalMediaForHold() {
        suspendLocalMediaForHoldCalls += 1
        // Mirror the real engine: hold RELEASES the mic + camera tracks.
        hasLocalAudioTrack = false
        hasLocalVideoTrack = false
        // Mirror the real engine: hold deafens remote playout. This also sets the
        // sticky flag so slots created later inherit the deafen.
        setRemotePlaybackEnabled(false)
    }

    func resumeLocalMediaFromHold(audioEnabled: Bool, videoMode: LocalCameraMode?) {
        resumeLocalMediaFromHoldCalls.append((audioEnabled: audioEnabled, videoMode: videoMode))
        // Mirror the real engine: resume reacquires per DESIRED intent only — a
        // muted resume leaves the audio track nil; a camera-off resume leaves the
        // video track nil. This is exactly the state the P5 toggle fix repairs.
        if audioEnabled { hasLocalAudioTrack = true }
        if videoMode != nil { hasLocalVideoTrack = true }
        // Mirror the real engine: resume re-enables remote playout.
        setRemotePlaybackEnabled(true)
    }

    /// Sticky deafen state, mirroring the real engine: held sessions set this
    /// `false` and a slot created AFTER hold must inherit it. Defaults enabled.
    private(set) var remotePlaybackEnabled = true
    func setRemotePlaybackEnabled(_ enabled: Bool) {
        remotePlaybackEnabled = enabled
        setRemotePlaybackEnabledCalls.append(enabled)
        fakeSlots.values.forEach { $0.setRemotePlaybackEnabled(enabled) }
    }

    func detachRenderersForHold() {
        detachRenderersForHoldCalls += 1
    }

    @discardableResult
    func toggleAudio(_ enabled: Bool) -> Bool {
        toggleAudioCalls.append(enabled)
        // P5 ensure-track: enabling with no audio track recreates it (resume-then-
        // unmute repair). When a track already exists this recreates NOTHING — the
        // single-call mute/unmute regression case. Returns the effective,
        // track-backed state.
        if enabled, !hasLocalAudioTrack {
            hasLocalAudioTrack = true
            toggleAudioTrackRecreations += 1
        }
        return enabled && hasLocalAudioTrack
    }

    func restartAudioUnit() {
        restartAudioUnitCalls += 1
    }

    @discardableResult
    func toggleVideo(_ enabled: Bool) -> Bool {
        toggleVideoCalls.append(enabled)
        // P5 ensure-track: enabling with no video track recreates it (resume-then-
        // video-on repair). When a track already exists this recreates NOTHING.
        if enabled, !hasLocalVideoTrack {
            hasLocalVideoTrack = true
            toggleVideoTrackRecreations += 1
        }
        return enabled && hasLocalVideoTrack
    }

    func flipCamera() {}
    func setHdVideoExperimentalEnabled(_ enabled: Bool) {}
    @discardableResult func toggleFlashlight() -> Bool { false }
    private(set) var startScreenShareCalls = 0
    private(set) var stopScreenShareCalls = 0
    /// Drives the result delivered to `startScreenShare`'s sync return AND the
    /// `onComplete` callback, so session-level tests can exercise both confirmed
    /// start (content_state broadcast) and cancellation/failure branches.
    var startScreenShareResult = false
    /// When true, model the broadcast PENDING window: return `true`
    /// (request accepted, reader listening) WITHOUT firing `onComplete`, so the
    /// share is not yet confirmed. A subsequent `stopScreenShare` then fires
    /// `onScreenShareStopped` (cancel the pending start) like the real controller.
    var deferStartScreenShareCompletion = false
    private var deferredStartScreenShareCompletion: ((Bool) -> Void)?
    func startScreenShare(onComplete: ((Bool) -> Void)?) -> Bool {
        startScreenShareCalls += 1
        if deferStartScreenShareCompletion {
            deferredStartScreenShareCompletion = onComplete
            return true
        }
        onComplete?(startScreenShareResult)
        return startScreenShareResult
    }
    func completeDeferredStartScreenShare(started: Bool) {
        let completion = deferredStartScreenShareCompletion
        deferredStartScreenShareCompletion = nil
        completion?(started)
    }
    func stopScreenShare() -> Bool {
        stopScreenShareCalls += 1
        // Mirror the real controller: drive the session's screen-share-stopped
        // callback so the session-side stop signaling path runs in tests.
        onScreenShareStopped?()
        return true
    }
    private(set) var adjustCaptureZoomCalls: [CGFloat] = []
    var adjustCaptureZoomResult: Double? = 1.25
    @discardableResult func adjustCaptureZoom(by scaleDelta: CGFloat) -> Double? {
        adjustCaptureZoomCalls.append(scaleDelta)
        return adjustCaptureZoomResult
    }
    @discardableResult func resetCaptureZoom() -> Double { 1.0 }

    func setIceServers(_ servers: [IceServerConfig]) {
        _iceServers = servers
        iceServersSet = true
        fakeSlots.values.forEach { $0.setIceServers(servers) }
    }

    func hasIceServers() -> Bool { _iceServers != nil }

    /// Per-cid record of the independent-content gate + offer-owner resolution
    /// each slot was created with, so routing tests can assert on them.
    private(set) var createdSlotSupportsIndependentContentVideo: [String: Bool] = [:]
    private(set) var createdSlotIsOfferOwner: [String: Bool] = [:]
    func createSlot(
        remoteCid: String,
        onLocalIceCandidate: @escaping (String, IceCandidatePayload) -> Void,
        onRemoteVideoTrack: @escaping (String, AnyObject?) -> Void,
        onConnectionStateChange: @escaping (String, String) -> Void,
        onIceConnectionStateChange: @escaping (String, String) -> Void,
        onSignalingStateChange: @escaping (String, String) -> Void,
        onRenegotiationNeeded: @escaping (String) -> Void,
        supportsIndependentContentVideo: Bool,
        isOfferOwner: @escaping () -> Bool
    ) -> (any PeerConnectionSlotProtocol)? {
        createdSlotCids.append(remoteCid)
        createdSlotSupportsIndependentContentVideo[remoteCid] = supportsIndependentContentVideo
        createdSlotIsOfferOwner[remoteCid] = isOfferOwner()
        let slot = FakePeerConnectionSlot(
            remoteCid: remoteCid,
            supportsIndependentContentVideo: supportsIndependentContentVideo,
            onConnectionStateChange: onConnectionStateChange,
            onIceConnectionStateChange: onIceConnectionStateChange,
            onSignalingStateChange: onSignalingStateChange,
            onRenegotiationNeeded: onRenegotiationNeeded
        )
        if failNextCreatedSlotRemoteOffer {
            slot.failNextRemoteOffer = true
            failNextCreatedSlotRemoteOffer = false
        }
        fakeSlots[remoteCid] = slot
        // Sticky deafen: a slot created while the session is held inherits the
        // disabled remote playback (mirrors WebRtcEngine.createSlot).
        if !remotePlaybackEnabled {
            slot.setRemotePlaybackEnabled(false)
        }
        if let iceServers = _iceServers {
            slot.setIceServers(iceServers)
        }
        return slot
    }

    func removeSlot(_ slot: any PeerConnectionSlotProtocol) {
        removedSlots.append(slot)
    }

    private(set) var attachLocalRendererCalls: [AnyObject] = []
    private(set) var detachLocalRendererCalls: [AnyObject] = []
    func attachLocalRenderer(_ renderer: AnyObject) {
        attachLocalRendererCalls.append(renderer)
    }
    func detachLocalRenderer(_ renderer: AnyObject) {
        detachLocalRendererCalls.append(renderer)
    }

    private(set) var attachLocalContentRendererCalls: [AnyObject] = []
    private(set) var detachLocalContentRendererCalls: [AnyObject] = []
    func attachLocalContentRenderer(_ renderer: AnyObject) {
        attachLocalContentRendererCalls.append(renderer)
    }
    func detachLocalContentRenderer(_ renderer: AnyObject) {
        detachLocalContentRendererCalls.append(renderer)
    }

    func setOnCameraFacingChanged(_ handler: @escaping (Bool) -> Void) {
        onCameraFacingChanged = handler
    }
    func setOnCameraModeChanged(_ handler: @escaping (LocalCameraMode) -> Void) {
        onCameraModeChanged = handler
    }
    func setOnFlashlightStateChanged(_ handler: @escaping (Bool, Bool) -> Void) {
        onFlashlightStateChanged = handler
    }
    func setOnScreenShareStopped(_ handler: @escaping () -> Void) {
        onScreenShareStopped = handler
    }
    func setOnZoomFactorChanged(_ handler: @escaping (Double) -> Void) {
        onZoomFactorChanged = handler
    }
    func setOnFeatureDegradation(_ handler: @escaping (FeatureDegradationState) -> Void) {
        onFeatureDegradation = handler
    }

    var nextLocalAudioLevel: Float?
    func collectLocalAudioLevel(_ onComplete: @escaping @Sendable (Float?) -> Void) {
        onComplete(nextLocalAudioLevel)
    }
}
